From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Todd Poynor <toddpoynor@google.com>
Date: Wed, 20 Mar 2013 15:40:46 -0700
Subject: [PATCH 3/4] cpufreq: interactive: fix crash on error paths in
 get_tokenized_data
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Use separate variable for error code, free proper pointer.

Change-Id: Ia83cccb195997789ac6afbf5b8761f7b278196d6
Reported-by: Arve Hjønnevåg <arve@android.com>
Signed-off-by: Todd Poynor <toddpoynor@google.com>

(cherry picked from commit cad1684410541be5363d4af9ae8f3f35683a4dd4 in AOSP kernel/common android-3.4 (now deprecated/android-3.4))
[OpenMyGlass: follow-up fix for the get_tokenized_data() added by the
 above_hispeed_delay backport; applies unchanged, only line offsets differ]
---
 drivers/cpufreq/cpufreq_interactive.c | 17 ++++++-----------
 1 file changed, 6 insertions(+), 11 deletions(-)

diff --git a/drivers/cpufreq/cpufreq_interactive.c b/drivers/cpufreq/cpufreq_interactive.c
index 14f9be9..c76c289 100644
--- a/drivers/cpufreq/cpufreq_interactive.c
+++ b/drivers/cpufreq/cpufreq_interactive.c
@@ -652,29 +652,26 @@ static unsigned int *get_tokenized_data(const char *buf, int *num_tokens)
 	int i;
 	int ntokens = 1;
 	unsigned int *tokenized_data;
+	int err = -EINVAL;
 
 	cp = buf;
 	while ((cp = strpbrk(cp + 1, " :")))
 		ntokens++;
 
-	if (!(ntokens & 0x1)) {
-		tokenized_data = ERR_PTR(-EINVAL);
+	if (!(ntokens & 0x1))
 		goto err;
-	}
 
 	tokenized_data = kmalloc(ntokens * sizeof(unsigned int), GFP_KERNEL);
 	if (!tokenized_data) {
-		tokenized_data = ERR_PTR(-ENOMEM);
+		err = -ENOMEM;
 		goto err;
 	}
 
 	cp = buf;
 	i = 0;
 	while (i < ntokens) {
-		if (sscanf(cp, "%u", &tokenized_data[i++]) != 1) {
-			tokenized_data = ERR_PTR(-EINVAL);
+		if (sscanf(cp, "%u", &tokenized_data[i++]) != 1)
 			goto err_kfree;
-		}
 
 		cp = strpbrk(cp, " :");
 		if (!cp)
@@ -682,10 +679,8 @@ static unsigned int *get_tokenized_data(const char *buf, int *num_tokens)
 		cp++;
 	}
 
-	if (i != ntokens) {
-		tokenized_data = ERR_PTR(-EINVAL);
+	if (i != ntokens)
 		goto err_kfree;
-	}
 
 	*num_tokens = ntokens;
 	return tokenized_data;
@@ -693,7 +688,7 @@ static unsigned int *get_tokenized_data(const char *buf, int *num_tokens)
 err_kfree:
 	kfree(tokenized_data);
 err:
-	return tokenized_data;
+	return ERR_PTR(err);
 }
 
 static ssize_t show_target_loads(
